### PR TITLE
Canvas: Connection original persistence check

### LINKED ISSUE #86379: Canvas: Connection source target persistence check on load

### PR DESCRIPTION
**Pull request by drew08t**
For already created connections, prior to storing source and target original coordinates, undesirable behavior occurred when moving elements. This checks that those original coordinates exist, and if they don't it creates them based on the current source and target coordinates.

Before:
Apr-17-2024 21-49-53

After:
Apr-17-2024 21-49-05

Fixes #86379

### CODE REVIEW

**Comment by drew08t** [comment]
I am seeing the following warning in the console after making this change:

Cannot update during an existing state transition (such as within `render`). Render methods should be a pure function of props and state.

File: public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx [hunk 1]
```diff
@@ -146,6 +146,18 @@ export const ConnectionSVG = ({
...
+            const currentConnections = [...source.options.connections];
+            // Check for original state
+            if (!currentConnections[index].sourceOriginal || !currentConnections[index].targetOriginal) {
+              currentConnections[index] = {
+                ...currentConnections[index],
+                sourceOriginal: { x: x1, y: y1 },
+                targetOriginal: { x: x2, y: y2 },
+              };
+              // Update model with populated originals
+              source.onChange({ ...source.options, connections: currentConnections });
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:159]: I am wondering if this type of one time logic would make better sense as migration code vs existing inside the render logic of connections indefinitely 🤔

Comment (drew08t) [reply public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:159]: Yeah, great suggestion! Let me look into moving it.

**Comment by drew08t** [comment]
> > I am seeing the following warning in the console after making this change:
> > Cannot update during an existing state transition (such as within `render`). Render methods should be a pure function of props and state.
> 
> Huh... Odd. I don't see you updating state in any new way anywhere... 🤔

I think it will be a moot point anyway since I intend to move this to migration anyway.

**Review by baldm0mma — APPROVED** [review]
I think after the migration code is moved, you're G2G!

**Review by codeincarnate — APPROVED** [review]
Agreed with moving the one-time logic but otherwise looks good to me!

File: public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx (hunk 1 above)
Comment (drew08t) [inline public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:159]: This onChange call is causing the warning.

Comment (drew08t) [reply public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:159]: I'm still unable to figure out how to access scene.connections.state from inside of migration. Open to suggestions here.

Comment (baldm0mma) [reply public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:159]: I don't suppose it's as easy as building a helper function that calls the onChange, then calling he helper function inside the render? Little bob an' weave? 😂

Comment (drew08t) [reply public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:159]: Good thought, but that didn't work.

Comment (nmarrs) [reply public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx]: Hm I looked more into this and yeah it seems like sticking this logic in a migration is not feasible given the data is based on runtime html dimensions and this logic run before the canvas panel is rendered. For future reference the properties of connections are accessible / changeable in migrations, but not in cases where the values are driven by runtime DOM placement.

```
const root = panel.options?.root;
if (root?.elements) {
  for (const element of root.elements) {
    if (element.connections) {
        for (const connection of element.connections) {
                // Logic here modifying connection properties
        }
      }
    }
}
```

Comment (nmarrs) [reply public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx]: It looks like we can avoid using this `onChange` call all together and just modify the state directly and avoid this issue, with this I think we can move forward and merge this PR as solution is "good enough" 😅

**Review by nmarrs — APPROVED** [review]
Tried out the migration approach and was unsuccessful due to not having access to runtime DOM given source / target original properties are relative DOM coords 😬 

Made slight modification, modifying the connection state in place to avoid the need to call `onChange` within the render and I think in its current state it is "good enough" merge and carry on 🚢 

Here is a test dashboard you can use to test / verify
[Old connection repro case-1714337469663.json](https://github.com/grafana/grafana/files/15143733/Old.connection.repro.case-1714337469663.json)

**Comment by drew08t** [comment]
> Tried out the migration approach and was unsuccessful due to not having access to runtime DOM given source / target original properties are relative DOM coords 😬
> 
> Made slight modification, modifying the connection state in place to avoid the need to call `onChange` within the render and I think in its current state it is "good enough" merge and carry on 🚢
> 
> Here is a test dashboard you can use to test / verify [Old connection repro case-1714337469663.json](https://github.com/grafana/grafana/files/15143733/Old.connection.repro.case-1714337469663.json)

Thanks for digging into this! I tested your approach and it works for me: no warning and the save model is updated appropriately.

**Pull request merged**